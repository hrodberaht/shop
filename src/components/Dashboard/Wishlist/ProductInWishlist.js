import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AddToCartButton from '../../shared/AddToCartButton';

export default class ProductInWishlist extends Component {
  handleClick = () => {
    const {
      remove,
      product: { productId },
    } = this.props;
    remove(productId);
  };

  render() {
    const {
      product,
      product: {
        name, price, pcsOrder, totalPrice,
      },
      handleClickToCart,
    } = this.props;
    return (
      <tr>
        <td>{name}</td>
        <td>{price}</td>
        <td>{pcsOrder}</td>
        <td>
          <span>$</span>
          {totalPrice}
        </td>
        <td>
          <button type="button" className="btn btn-danger" onClick={this.handleClick}>
            X
          </button>
        </td>
        <td>
          <AddToCartButton
            className="btn btn-primary"
            product={product}
            handleClickToCart={handleClickToCart}
          />
        </td>
      </tr>
    );
  }
}

ProductInWishlist.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string,
    type: PropTypes.string,
    price: PropTypes.number,
  }).isRequired,
  remove: PropTypes.func.isRequired,
  handleClickToCart: PropTypes.func.isRequired,
};
