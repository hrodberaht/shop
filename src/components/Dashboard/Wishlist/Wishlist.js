import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchWishlist, removeProductWishlist } from '../../../store/wishlist/actionCerators';
import { getAuthToken, getAuthUserId } from '../../../store/authenticate/selectors';
import { getWishlistLoaded, getWishlistProducts } from '../../../store/wishlist/selectors';
import ProductInWishlist from './ProductInWishlist';
import { addProductToCart, updateProductInCart } from '../../../store/cart/actionCreators';
import { getIdsProductsInCart } from '../../../store/cart/selectors';

export class Wishlist extends Component {
  componentDidMount() {
    const { getWishlist, userId, token } = this.props;
    getWishlist(userId, token);
  }

  remove = (productId) => {
    const { removeProd, userId, token } = this.props;
    removeProd(productId, userId, token);
  };

  handleClickToCart = (product) => {
    const { addProduct, updateProduct, idsProductsInCart } = this.props;
    if (idsProductsInCart.includes(product.productId)) {
      updateProduct(product);
    } else {
      addProduct(product);
    }
    this.remove(product.productId);
  };

  render() {
    const { loaded, products } = this.props;
    if (!loaded) return <h4>Loading</h4>;
    return (
      <div className="wishlist">
        <h3>Wish list:</h3>
        <table>
          <tbody>
            <tr>
              <th>Name:</th>
              <th>Price:</th>
              <th>Pcs:</th>
              <th>Total price:</th>
              <th>Delete:</th>
              <th>Add to cart:</th>
            </tr>
            {products.map(prod => (
              <ProductInWishlist
                product={prod}
                remove={this.remove}
                key={prod.productId}
                handleClickToCart={this.handleClickToCart}
              />
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  token: getAuthToken(state),
  userId: getAuthUserId(state),
  loaded: getWishlistLoaded(state),
  products: getWishlistProducts(state),
  idsProductsInCart: getIdsProductsInCart(state),
});

export default connect(
  mapStateToProps,
  {
    getWishlist: fetchWishlist,
    removeProd: removeProductWishlist,
    addProduct: addProductToCart,
    updateProduct: updateProductInCart,
  },
)(Wishlist);

Wishlist.propTypes = {
  getWishlist: PropTypes.func,
  userId: PropTypes.string,
  token: PropTypes.string.isRequired,
  loaded: PropTypes.bool.isRequired,
  products: PropTypes.arrayOf(PropTypes.objectOf),
  removeProd: PropTypes.func.isRequired,
  addProduct: PropTypes.func.isRequired,
  updateProduct: PropTypes.func.isRequired,
  idsProductsInCart: PropTypes.arrayOf(PropTypes.string).isRequired,
};

Wishlist.defaultProps = {
  products: null,
  getWishlist: null,
  userId: null,
};
