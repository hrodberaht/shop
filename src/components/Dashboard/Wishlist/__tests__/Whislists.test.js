import { Wishlist } from '../Wishlist';

const defaultProps = {
  products: [
    {
      productId: 'product-1',
      type: 'Printer',
      name: 'ZX3',
      price: 400,
      inStock: '30',
    },
    {
      productId: 'product-2',
      type: 'Printer',
      name: 'ZX5',
      price: 600,
      inStock: '5',
    },
  ],
  userId: '1234',
  getWishlist: jest.fn(),
  token: '12345',
  loaded: false,
  removeProd: jest.fn(),
  addProduct: jest.fn(),
  updateProduct: jest.fn(),
  idsProductsInCart: ['product-1'],
};

const setup = buildSetup(Wishlist, defaultProps);

describe('<Wishlist />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render products list if loaded true', () => {
    const { wrapper } = setup();
    wrapper.setProps({ loaded: true });
    expect(wrapper).toMatchSnapshot();
  });
  it('should call handleFetch with token', () => {
    const {
      wrapper,
      props: { userId, token },
    } = setup();
    expect(wrapper.instance().props.getWishlist).toHaveBeenCalledWith(userId, token);
  });
  it('should call removeProd if remove is call', () => {
    const {
      wrapper,
      props: { products, userId, token },
    } = setup();
    const { productId } = products[0];
    wrapper.instance().remove(productId);
    expect(wrapper.instance().props.removeProd).toHaveBeenCalledWith(productId, userId, token);
  });
  it('should call addProduct when products is not in cart and not call updateProduct', () => {
    const {
      wrapper,
      props: { products },
    } = setup();
    const instance = wrapper.instance();
    instance.handleClickToCart(products[1]);
    expect(wrapper.instance().props.updateProduct).not.toHaveBeenCalled();
    expect(wrapper.instance().props.addProduct).toHaveBeenCalledWith(products[1]);
  });

  it('should call update updateProduct when products is in cart and not call addProduct', () => {
    const {
      wrapper,
      props: { products },
    } = setup();
    const instance = wrapper.instance();
    instance.props.addProduct.mockClear();
    instance.handleClickToCart(products[0]);
    expect(wrapper.instance().props.updateProduct).toHaveBeenCalledWith(products[0]);
    expect(wrapper.instance().props.addProduct).not.toHaveBeenCalled();
  });
});
