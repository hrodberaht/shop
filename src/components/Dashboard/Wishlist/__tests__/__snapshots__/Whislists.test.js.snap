// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Wishlist /> should render products list if loaded true 1`] = `
<div
  className="wishlist"
>
  <h3>
    Wish list:
  </h3>
  <table>
    <tbody>
      <tr>
        <th>
          Name:
        </th>
        <th>
          Price:
        </th>
        <th>
          Pcs:
        </th>
        <th>
          Total price:
        </th>
        <th>
          Delete:
        </th>
        <th>
          Add to cart:
        </th>
      </tr>
      <ProductInWishlist
        handleClickToCart={[Function]}
        key="product-1"
        product={
          Object {
            "inStock": "30",
            "name": "ZX3",
            "price": 400,
            "productId": "product-1",
            "type": "Printer",
          }
        }
        remove={[Function]}
      />
      <ProductInWishlist
        handleClickToCart={[Function]}
        key="product-2"
        product={
          Object {
            "inStock": "5",
            "name": "ZX5",
            "price": 600,
            "productId": "product-2",
            "type": "Printer",
          }
        }
        remove={[Function]}
      />
    </tbody>
  </table>
</div>
`;

exports[`<Wishlist /> should render without crash 1`] = `
<h4>
  Loading
</h4>
`;
